### ISSUE TITLE
[BUG]ActiveLimitFilter 在消费者并发调用时不能真正限制并发数if (active >= max)与 RpcStatus.beginCount(url, methodName);操作不是原子的

### ISSUE THREAD

**Issue by wj1369884616** [apache/incubator-dubbo#2328]
```java 
        URL url = invoker.getUrl();
        String methodName = invocation.getMethodName();
        int max = invoker.getUrl().getMethodParameter(methodName, Constants.ACTIVES_KEY, 0);
        RpcStatus count = RpcStatus.getStatus(invoker.getUrl(), invocation.getMethodName());
        if (max > 0) {
            long timeout = invoker.getUrl().getMethodParameter(invocation.getMethodName(), Constants.TIMEOUT_KEY, 0);
            long start = System.currentTimeMillis();
            long remain = timeout;
            int active = count.getActive();
            if (active >= max) {
                synchronized (count) {
                    while ((active = count.getActive()) >= max) {
                        try {
                            count.wait(remain);
                        } catch (InterruptedException e) {
                        }
                        long elapsed = System.currentTimeMillis() - start;
                        remain = timeout - elapsed;
                        if (remain <= 0) {
                            throw new RpcException("Waiting concurrent invoke timeout in client-side for service:  "
                                    + invoker.getInterface().getName() + ", method: "
                                    + invocation.getMethodName() + ", elapsed: " + elapsed
                                    + ", timeout: " + timeout + ". concurrent invokes: " + active
                                    + ". max concurrent invoke limit: " + max);
                        }
                    }
                }
            }
        }
        try {
            long begin = System.currentTimeMillis();
            RpcStatus.beginCount(url, methodName);
            try {
                Result result = invoker.invoke(invocation);
                RpcStatus.endCount(url, methodName, System.currentTimeMillis() - begin, true);
                return result;
            } catch (RuntimeException t) {
                RpcStatus.endCount(url, methodName, System.currentTimeMillis() - begin, false);
                throw t;
            }
        } finally {
            if (max > 0) {
                synchronized (count) {
                    count.notify();
                }
            }
        }
    }
```

**Comment by ralf0131** [comment]
Hi, 

Would you please send a pull request?

**Comment by ningyu1** [comment]
repaired PR: #2437

**Comment by wj1369884616** [comment]
ok,I will be send a pull request  after two days。

**Comment by wj1369884616** [comment]
repaired PR:#2445

**Comment by diecui1202** [comment]
@ningyu1 @wj1369884616 I find there has 2 PRs, #2445 #2437 

could u two take more discussion about this issue, and merge them to one PR. thanks.

**Comment by ningyu1** [comment]
@diecui1202 @wj1369884616 
My 'pr #2437' is a relatively small change in the old structure to ensure atomicity.
His 'pr #2445' is replacing the while with the semaphore, which I think is good.
I think using the semaphore is better than my repair.
I don't know what you think?

**Comment by wj1369884616** [comment]
@ningyu1 @diecui1202 Using the semaphore, the thread will not block when the number of threads is less than  or equal to the maximum number of threads, so I think the performance is better.

**Comment by ningyu1** [comment]
@wj1369884616 Consent

**Comment by beiwei30** [comment]
it's been fixed with 69e72041943f247360249ba191ac1217217f43a5

**Comment by scxwhite** [comment]
> it's been fixed with [69e7204](https://github.com/apache/incubator-dubbo/commit/69e72041943f247360249ba191ac1217217f43a5) and [e6e7ea8](https://github.com/apache/incubator-dubbo/commit/e6e7ea8b12d952a9ccb53e73ace18231effbf767)

I do not thind it is been fixed . In  `RpcStatus.beginCount `method, `if(methodStatus.active.incrementAndGet() > max)` will allow more request(> max). because `incrementAndGet` is not atomic calc. you can run this unit-test.


`public void testAtomicInteger() throws InterruptedException {
        AtomicInteger integer = new AtomicInteger();

        ExecutorService threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());

        int size = 10000;
        CountDownLatch latch = new CountDownLatch(size);
        HashSet<Integer> set = new HashSet<>(size);
        for (int i = 0; i < size; i++) {
            threadPool.execute(() -> {
                set.add(integer.incrementAndGet());
                latch.countDown();
            });
        }
        latch.await();
        System.out.println("the size is :" + set.size());
    }`
the `set.size()` will less than  `10000` in more case
